internal/evaluator/rego.go

```diff
@@ -0,0 +1,134 @@
+// Copyright The Enterprise Contract Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+// IMPORTANT: The rego functions in this file never return an error. Instead, they return no value
+// when an error is encountered. If they did return an error, opa would exit abruptly and it would
+// not produce a report of which policy rules succeeded/failed.
+
+package evaluator
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"io"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	gcr "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/open-policy-agent/opa/types"
+	log "github.com/sirupsen/logrus"
+)
+
+type key string
+
+const remoteLayerKey key = "ec.evaluator.remote.layer"
+
+type remoteLayerFn func(ref name.Digest, options ...remote.Option) (gcr.Layer, error)
+
+func remoteLayer(ctx context.Context, ref name.Digest, options ...remote.Option) (gcr.Layer, error) {
+	fn, ok := ctx.Value(remoteLayerKey).(remoteLayerFn)
+	if ok && fn != nil {
+		return fn(ref, options...)
+	}
+	return remote.Layer(ref, options...)
```

Comment:
Could we use cache here? Like what we do in `internal/fetchers/oci/client.go`?

Reply:
+1 done